sound/soc/sof/ipc4-pcm.c

```diff
@@ -32,6 +34,174 @@ int sof_ipc4_set_pipeline_state(struct snd_sof_dev *sdev, u32 id, u32 state)
+		dev_err(sdev->dev, "failed to find host copier %p or dai copier%p\n",
+			host_copier, dai_copier);
+		return;
+	}
+
+	if (host_copier->data.gtw_cfg.node_id >= SOF_IPC4_MAX_PIPELINE_REG_SLOTS) {
+		dev_err(sdev->dev, "invalid node_id 0x%x\n", host_copier->data.gtw_cfg.node_id);
+		return;
+	}
+
+	fw_reg = (struct sof_ipc4_fw_registers *)(sdev->fw_info_box.offset +
+						sdev->bar[sdev->mailbox_bar]);
+	node_id = SOF_IPC4_NODE_INDEX(host_copier->data.gtw_cfg.node_id);
+	ppl_reg = &fw_reg->pipeline_regs[node_id];
+	/* it is unlikely that this offset is still invalid after pipeline is started 10ms later */
+	if (ppl_reg->stream_start_offset == SOF_IPC4_INVALID_STREAM_POSITION) {
+		dev_err(sdev->dev, "failed to get stream start offset\n");
+		return;
+	}
+	dev_dbg(sdev->dev, "stream_start_offset position %lld\n",  ppl_reg->stream_start_offset);
+
+	for (i = 0; i < SOF_IPC4_MAX_LLP_GPDMA_READING_SLOTS; i++) {
+		if (fw_reg->llp_gpdma_reading_slots[i].node_id ==
+			dai_copier->data.gtw_cfg.node_id) {
+			llp_slot = fw_reg->llp_gpdma_reading_slots + i;
+			break;
+		}
+	}
+	/* if it is not found, check aggregated sdw slot */
+	for (i = 0; !llp_slot && i < SOF_IPC4_MAX_LLP_SNDW_READING_SLOTS - 1; i++) {
+		if (fw_reg->llp_sndw_reading_slots[i].node_id == dai_copier->data.gtw_cfg.node_id) {
+			llp_slot = fw_reg->llp_sndw_reading_slots + i;
+			break;
+		}
+	}
+
+	if (!llp_slot) {
+		dev_err(sdev->dev, "failed to find node_id 0x%x\n",
+			dai_copier->data.gtw_cfg.node_id);
+		return;
+	}
+
+	spcm->llp_slot = llp_slot;
+
+	pos_pre = llp_slot->reading.llp_l + ((u64)llp_slot->reading.llp_u << 32);
+	wclk_pre = llp_slot->reading.wclk_l + ((u64)llp_slot->reading.wclk_u << 32);
+
+	/* wait for more than 1ms to get llp updated by fw */
+	msleep(1.1);
+	pos_current = llp_slot->reading.llp_l + ((u64)llp_slot->reading.llp_u << 32);
+	wclk_current = llp_slot->reading.wclk_l + ((u64)llp_slot->reading.wclk_u << 32);
+
+	if (!(pos_current - pos_pre)) {
+		dev_err(sdev->dev, "invalid position\n");
+		return;
+	}
+
+	stream_start_position = ppl_reg->stream_start_offset;
+	dai_sample_size = (dai_copier->data.out_format.bit_depth >> 3) *
+		SOF_IPC4_AUDIO_FORMAT_CFG_CHANNELS_COUNT(dai_copier->data.out_format.fmt_cfg);
+	/* convert offset to sample count */
+	do_div(stream_start_position, dai_sample_size);
+
+	/*
+	 * ratio = (wclk_current - wclk_pre) / pos_current - pos_pre
+	 */
+	ratio = wclk_current - wclk_pre;
+	do_div(ratio, pos_current - pos_pre);
+	spcm->stream_start_wclk = wclk_current - ratio * (pos_current - stream_start_position);
+
+	dev_dbg(sdev->dev, "current clk %lld fw start time wclk = %lld\n",
+		wclk_current, spcm->stream_start_wclk);
+
+	if (!spcm->accuracy) {
+		wclk = ratio * dai_copier->data.out_format.sampling_frequency;
+		dev_dbg(sdev->dev, "current wclk = %lld\n", wclk);
+
+		sof_ipc4_timecounter_init(sdev, spcm, wclk);
+	}
+
+	/* audio start time is 0 since trigger */
+	timecounter_init(&spcm->tc, &spcm->cc, 0);
```

Comment:
I couldn't figure out why the timecounters are initialized here *AFTER* the stream started. Please add comments or revisit.

Reply:
sure. The problem is that we don't know wclk and stream_start_offset before stram started,  so we need to calibrate them after pipeline start.

Reply:
in legacy HDA code, timecounter is also initialized after snd_hdac_stream_start 
````c
	snd_pcm_group_for_each_entry(s, substream) {
		if (s->pcm->card != substream->pcm->card)
			continue;
		azx_dev = get_azx_dev(s);
		if (start) {
			azx_dev->insufficient = 1;
			snd_hdac_stream_start(azx_stream(azx_dev), true);
		} else {
			snd_hdac_stream_stop(azx_stream(azx_dev));
		}
	}
	spin_unlock(&bus->reg_lock);

	snd_hdac_stream_sync(hstr, start, sbits);

	spin_lock(&bus->reg_lock);
	/* reset SYNC bits */
	snd_hdac_stream_sync_trigger(hstr, false, sbits, sync_reg);
	if (start)
		snd_hdac_stream_timecounter_init(hstr, sbits);
````